Fix Reinforcements.ReinforceWithTransport for planes

abcdefg30 opened the pull request:
Changes
![image](http://imgur.com/EkYfgVS.gif)
to
![image](http://imgur.com/geb68sZ.gif)

abcdefg30 committed:
Add lua support for landing on a certain position

abcdefg30 committed:
Fix Reinforcements.ReinforceWithTransport for planes

pchote commented:
Isn't this kind of bogus?  Aircraft aren't supposed to be able to land on the ground.

GraionDilach commented:
Mission aircraft could be. This looks like something allowing to simulate ProductionAirdrop to me without a dummy producer actor.

abcdefg30 commented:
> Isn't this kind of bogus? Aircraft aren't supposed to be able to land on the ground.

Checks for `LandableTerrainTypes` could be added, but I think that's an unnecessary restriction/ complication. You can already create actors at "bogus" (unpathable) locations (e.g. rilfe men on water tiles). That is simply a scripting error then.

abcdefg30 commented:
I'm also not sure if the production airdrop mentioned by @GraionDilach is possible when adding checks for `LandableTerrainTypes`. We'll at least need that for some d2k missions.

reaperrr commented:
> Isn't this kind of bogus? Aircraft aren't supposed to be able to land on the ground.

Maybe not in our mods, but at least for 3rd-party mods (any maybe some missions) this is a sensible feature, in my opinion.

obrakmann commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -125,11 +125,16 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
+			for (var i = 1; i < entryPath.Length - 1; i++)
 				Move(transport, entryPath[i]);
 
+			var lastPosition = entryPath.Last();
```
Can you rename this `destination` or `dropoffPoint` or something? `lastPosition` is confusing.

obrakmann commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -140,29 +145,44 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
+
+						var delta = transport.CenterPosition - transport.World.Map.CenterOfCell(lastPosition);
+						var sign = delta.LengthSquared >= landDistance * landDistance ? 1 : -1;
+						var positionBeforeLanding = lastPosition + sign * new CVec((int)horizontal, (int)vertical);
```
Something must be wrong with this. It causes the noticeable drop right before touchdown. If I add 1024 to the `landDistance`, the landing is smooth as expected, but I'm not sure if that's actually the part where the error is, and I forgot too much of my highschool math to understand what's happening here.

obrakmann reviewed (dismissed):
Looks otherwise ok

pchote commented:
@abcdefg30: Are you planning to fix this, or shall we close it?

abcdefg30 commented:
I plan to look into this the coming week.

reaperrr commented:
> I plan to look into this the coming week.

That obviously didn't work out, will you resume work on it soon or can this be closed for now?

abcdefg30 commented:
Sorry for the very late update. @obrakmann fixed those issues. I rewrote quite a bit of the original code and added some comments, so I hope it easier to understand now.

pchote commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -128,11 +128,16 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
+			var dropoffPoint = entryPath.Last();
 			if (actionFunc != null)
 			{
+				Move(transport, dropoffPoint);
```
If `entryPath` has one element then this will queue a `Move` to it that we don't want (it is excluded in the old code by starting the loop at index 1).

pchote commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -143,29 +148,60 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
+						var dropoffCenter = map.CenterOfCell(dropoffPoint);
+						var speed = aircraft.Info.Speed;
+
+						// Distance from the altitude to the ground
```
🚨 DUPLICATION ALERT 🚨 

Please adapt `ReturnToBase` to cover your use case, don't just copy all of its code here!

abcdefg30 commented:
This'll want to wait for #13785.

pchote commented:
Closing stale PR.  You know the boilerplate about reopening when ready by now 😄
